src/main.rs

```diff
@@ -100,6 +110,219 @@ fn get_username(
         })
 }
 
+/// Recursively creates all necessary parent directories for a given path,
+/// applying the specified numeric mode (e.g., 0o755) to each directory.
+fn create_dirs_with_mode(path: &Path, mode: u32) -> Result<()> {
+    if path.is_dir() {
+        return Ok(());
+    }
+    if let Some(parent) = path.parent() {
+        if !parent.as_os_str().is_empty() {
+            create_dirs_with_mode(parent, mode)?;
+        }
+    }
+
+    let mut builder = DirBuilder::new();
+    builder
+        .mode(mode)
+        .create(path)
+        .with_context(|| format!("Failed to create directory {:?}", path))?;
+    Ok(())
+}
+
+/// Applies a list of `WriteFiles` specifications, creating or overwriting each file
+/// according to its config. If `fail_provisioning_upon_failure` is `true` for a file
+/// and an error occurs, provisioning will fail. Otherwise, the error is logged and processing continues.
+///
+/// Returns a string describing the last successfully processed file (if any), or an error if one occurs
+/// with `fail_provisioning_upon_failure = true`.
+fn apply_write_files(
```

Comment:
I would recommend organizing these methods into a different module and not stuff them into main.rs. This will allow future improvements to be limited to the module and easier to review. Perhaps `write_file.rs` and `runcmd.rs`